src/filter.rs

```diff
@@ -0,0 +1,300 @@
+/// filter chain.
+///
+/// Modeled after git2/sys/filter.h in a streaming manner.
+pub trait StreamFilter {
+    /// Customized WriteStream implementation.
+    type S: WriteStream;
+
+    /// Create a new stream to handle new blob, should return a initialized
+    /// WriteStream.
+    fn open_stream(&self, source: FilterSource, next: RawStream)
+        -> Result<Self::S, Error>;
+}
+
+impl WriteStream for RawStream {
+    fn write(&mut self, buf: &[u8])
+        -> Result<(), Error> {
+        let code = unsafe {
+            ((*self.stream).write)((*self).stream,
+                buf.as_ptr() as *const c_char, buf.len())
+        };
+        call::try(code).and(Ok(()))
+    }
+
+    fn close(&mut self) -> Result<(), Error> {
+        let code = unsafe { ((*self.stream).close)(self.stream) };
+        call::try(code).and(Ok(()))
+    }
+}
+
+extern fn stream_write<T: WriteStream>(stream: *mut git_writestream,
+                       buf: *const c_char,
+                       len: size_t) -> c_int {
+    let ws = stream as *mut Stream<T>;
+    unsafe {
+        let input = slice::from_raw_parts(buf as *const u8, len);
+        match (*ws).handler.write(input) {
+            Ok(..) => raw::GIT_OK,
+            Err(_) => raw::GIT_ERROR,
+        }
+    }
+}
+
+extern fn stream_free<T: WriteStream>(stream: *mut git_writestream) -> c_int {
+    let _ = unsafe { Box::from_raw(stream as *mut Stream<T>) };
+    raw::GIT_OK
+}
+
+extern fn stream_close<T: WriteStream>(stream: *mut git_writestream) -> c_int {
+    let ws = stream as *mut Stream<T>;
+    unsafe {
+        match (*ws).handler.close() {
+            Ok(..) => raw::GIT_OK,
+            Err(_) => raw::GIT_ERROR,
+        }
+    }
+}
+
+extern fn filter_init(_filter: *mut git_filter) -> c_int {
+    raw::GIT_OK
+}
+
+extern fn filter_shutdown<T>(filter: *mut git_filter) {
+    let _ = unsafe { Box::from_raw(filter as *mut Filter<T>) };
+    // dropped
+}
+
+extern fn filter_check(_filter: *mut git_filter,
+                       _payload: *mut *mut c_void,
+                       _src: *const raw::git_filter_source,
+                       _attr_values: *const *mut c_char) -> c_int {
+    raw::GIT_OK
+}
+
+extern fn filter_stream_init<T: StreamFilter>(out: *mut *mut raw::git_writestream,
+                        filter: *mut git_filter,
+                        _payload: *mut *mut c_void,
+                        src: *const raw::git_filter_source,
+                        next: *mut raw::git_writestream) -> c_int
+{
+    let mode = unsafe { raw::git_filter_source_mode(src) };
+    let filter = filter as *mut Filter<T>;
+    let next = RawStream { stream: next };
+    let source = FilterSource {
+        to_worktree: (mode == raw::GIT_FILTER_TO_WORKTREE),
+        path: unsafe { util::bytes2path(CStr::from_ptr(raw::git_filter_source_path(src)).to_bytes())
+            .to_path_buf() }
+    };
+    let stream = unsafe { match (*filter).instance.open_stream(source, next) {
```

Comment:
A number of locations like this will need the panic protection around them to ensure that panics are handled.